mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -191,98 +403,74 @@ impl PeerManager {
 
         let sync_state = peer_info.get_mut(&peer).expect("missing peer sync state");
 
-        if let Some(batch) = sync_state.broadcast_info.sent_batches.remove(&batch_id) {
-            // update ACK counter
-            counters::SHARED_MEMPOOL_PENDING_BROADCASTS_COUNT
-                .with_label_values(&[&peer.peer_id().to_string()])
-                .dec();
-
+        if let Some(batch) = sync_state.broadcast_info.sent_batches.get(&batch_id) {
             // track broadcast roundtrip latency
-            if let Some(rtt) = timestamp.checked_duration_since(batch.timestamp) {
+            if let Ok(rtt) = timestamp.duration_since(batch.send_time) {
                 counters::SHARED_MEMPOOL_BROADCAST_RTT
                     .with_label_values(&[&peer.peer_id().to_string()])
                     .observe(rtt.as_secs_f64());
             }
 
-            // convert retry_txns from index within a batch to actual timeline ID of txn
-            let retry_timeline_ids = retry_txns
-                .iter()
-                .filter_map(|batch_index| batch.timeline_ids.get(*batch_index as usize).cloned())
-                .collect::<HashSet<_>>();
-            for timeline_id in batch.timeline_ids.into_iter() {
-                if retry_timeline_ids.contains(&timeline_id) {
-                    // add this retriable txn's timeline ID to peer info's retry_txns
-                    sync_state
-                        .broadcast_info
-                        .total_retry_txns
-                        .insert(timeline_id);
-                } else {
-                    // this txn was successfully ACK'ed for - aggressively remove from retry_txns
-                    sync_state
-                        .broadcast_info
-                        .total_retry_txns
-                        .remove(&timeline_id);
-                }
-            }
+            // track ACK counter
+            counters::SHARED_MEMPOOL_PENDING_BROADCASTS_COUNT
+                .with_label_values(&[&peer.peer_id().to_string()])
+                .dec();
+        } else {
+            error!(
+                "[mempool] received ACK for unknown batch id {:?} from {:?}",
+                batch_id, peer
+            );
+            return;
+        }
+
+        // For a fancier broadcasting strategy, we can only try to resend retriable txns
+        // versus resending the entire batch
+        // However, per-transaction batch analysis is not scalable CPU-wise/not worth the complexity,
+        // so for efficiency, we resend the entire batch if any of its txns are retriable
+        if retry_txns.is_empty() {
+            // remove broadcast from pending
+            sync_state.broadcast_info.sent_batches.remove(&batch_id);
+        } else {
+            sync_state
+                .broadcast_info
+                .sent_batches
+                .entry(batch_id)
+                .and_modify(|batch| batch.ack_status = AckStatus::Retry);
+        }
```

Comment:
rather than have a general batches state with two fields, I would recommend two Batches sets one called waiting and one called retry